### PR TITLE
Make it possible to calculate the storage root as often as you want

### PR DESCRIPTION
**Pull request by bkchr**
So, until now each Substrate based blockchain has calculated the storage
root once, at the end of the block. Now there is Frontier that wants to
calculate some intermediate storage root. However this failed on block
import. The problem with that was the extrinsics root. When building the
block we stored `Default::default()` as extrinsics root, because yeah,
we don't know the extrinsics root before finishing the block. At the end
this extrinsics root was then calculated. But on block import we passed
the already known extrinsics root. This was no problem, as we removed
this value at the end of the block. However when you all the storage
root in between, that changes the storage root between block building
and block import.

This pr changes this behavior. It removes the `ExtrinsicsRoot` storage
entry and also doesn't pass it anymore to `System::initialize`. By doing
it, we remove the difference in the storage and fix the storage root mismatch.

Fixes: https://github.com/paritytech/substrate/issues/7625

Cc @sorpaas @tgmichel

### CODE REVIEW

**Comment by bkchr** [comment]
Still misses some migration code to remove the `ExtrinsicsRoot` field from storage.

**Review by sorpaas — APPROVED** [review]

**Review by faith — APPROVED** [review]

**Comment by bkchr** [comment]
> Still misses some migration code to remove the `ExtrinsicsRoot` field from storage.

That was a dumb comment, this value was always removed at the end of a block. So, no need to remove it manually ;)

File: frame/executive/src/lib.rs [hunk 1]
```diff
@@ -286,13 +283,8 @@ where
...
 			n > System::BlockNumber::zero()
 			&& <frame_system::Module<System>>::block_hash(n - System::BlockNumber::one()) == *header.parent_hash(),
-			"Parent hash should be valid."
+			"Parent hash should be valid.",
 		);
-
-		// Check that transaction trie root represents the transactions.
-		let xts_root = extrinsics_root::<System::Hashing, _>(&block.extrinsics());
-		header.extrinsics_root().check_equal(&xts_root);
-		assert!(header.extrinsics_root() == &xts_root, "Transaction trie root must be valid.");
```
Comment (thiolliere) [inline frame/executive/src/lib.rs:295]: just to be sure, there wasn't need for this to be moved in `final_checks` except for avoiding duplicated code, isn't it ?

File: frame/executive/src/lib.rs [hunk 2]
```diff
@@ -357,15 +353,6 @@ where
...
-	fn apply_extrinsic_no_note(uxt: Block::Extrinsic) {
-		let l = uxt.encode().len();
-		match Self::apply_extrinsic_with_len(uxt, l, None) {
-			Ok(_) => (),
-			Err(e) => { let err: &'static str = e.into(); panic!(err) },
-		}
-	}
-
 	/// Actually apply an extrinsic given its `encoded_len`; this doesn't note its hash.
 	fn apply_extrinsic_with_len(
```
Comment (thiolliere) [inline frame/executive/src/lib.rs:357]: `to_note` being an `Option` seems useless now. But we can keep it as is.

**Review by thiolliere — APPROVED** [review]
looks good to me, the code looks also simpler now.

File: frame/system/src/lib.rs [hunk 3]
```diff
@@ -1137,9 +1132,6 @@ impl<T: Config> Module<T> {
 	/// Note what the extrinsic data of the current extrinsic index is. If this
 	/// is called, then ensure `derive_extrinsics` is also called before
```
[the remark this answers is not in the document]
Comment (thiolliere) [reply frame/system/src/lib.rs:1133]: but actually the documentation is a bit weird now I think

File: frame/executive/src/lib.rs (hunk 1 above)
Comment (bkchr) [reply frame/executive/src/lib.rs:295]: Removes duplicate code and work. We calculate the extrinsics root while finalizing the block. This means we don't need to calculate here again.

Comment (thiolliere) [reply frame/executive/src/lib.rs:295]: yes all clear

**Review by shawntabrizi — APPROVED** [review]

**Pull request merged**